app/src/main/java/android/support/v17/leanback/supportleanbackshowcase/app/media/ExoPlayerAdapter.java

```diff
@@ -0,0 +1,312 @@
+        boolean hadDisplay = mHasDisplay;
+        mHasDisplay = surfaceHolder != null;
+        if (hadDisplay == mHasDisplay) {
+            return;
+        }
+
+        mPlayer.setVideoSurfaceHolder(surfaceHolder);
+        if (mHasDisplay) {
+            if (mInitialized) {
+                getCallback().onPreparedStateChanged(ExoPlayerAdapter.this);
+            }
+        } else {
+            if (mInitialized) {
+                getCallback().onPreparedStateChanged(ExoPlayerAdapter.this);
+            }
+        }
+    }
+
+    @Override
+    public void setProgressUpdatingEnabled(final boolean enabled) {
+        mHandler.removeCallbacks(mRunnable);
+        if (!enabled) {
+            return;
+        }
+        mHandler.postDelayed(mRunnable, getUpdatePeriod());
+    }
+
+    int getUpdatePeriod() {
+        return 16;
+    }
+
+    @Override
+    public boolean isPlaying() {
+        boolean exoPlayerIsPlaying = mPlayer.getPlaybackState() == ExoPlayer.STATE_READY
+                && mPlayer.getPlayWhenReady();
+        return mInitialized && exoPlayerIsPlaying;
+    }
+
+    @Override
+    public long getDuration() {
+        return mInitialized ? mPlayer.getDuration() : -1;
+    }
+
+    @Override
+    public long getCurrentPosition() {
+        return mInitialized ? mPlayer.getCurrentPosition() : -1;
+    }
+
+
+    @Override
+    public void play() {
+        if (!mInitialized || isPlaying()) {
+            return;
+        }
+
+        mPlayer.setPlayWhenReady(true);
+        getCallback().onPlayStateChanged(ExoPlayerAdapter.this);
+        getCallback().onCurrentPositionChanged(ExoPlayerAdapter.this);
+    }
+
+    @Override
+    public void pause() {
+        if (isPlaying()) {
+            mPlayer.setPlayWhenReady(false);
+            getCallback().onPlayStateChanged(ExoPlayerAdapter.this);
+        }
+    }
+
+    @Override
+    public void seekTo(long newPosition) {
+        if (!mInitialized) {
+            return;
+        }
+        mPlayer.seekTo((int) newPosition);
+    }
+
+    @Override
+    public long getBufferedPosition() {
+        return mPlayer.getBufferedPosition();
+    }
+
+    /**
+     * Sets the media source of the player with a given URI.
+     *
+     * @return Returns <code>true</code> if uri represents a new media; <code>false</code>
+     * otherwise.
+     * @see ExoPlayer#prepare(MediaSource)
+     */
+    public boolean setDataSource(Uri uri) {
+        if (mMediaSourceUri != null ? mMediaSourceUri.equals(uri) : uri == null) {
+            return false;
+        }
+        mMediaSourceUri = uri;
+        prepareMediaForPlaying();
+        return true;
+    }
+
+    private void prepareMediaForPlaying() {
+        reset();
+        if (mMediaSourceUri != null) {
+            String userAgent = Util.getUserAgent(mContext, "ClassicalMusicQuiz");
+            MediaSource mediaSource = new ExtractorMediaSource(mMediaSourceUri,
+                    new DefaultDataSourceFactory(mContext, userAgent),
```

Comment:
Should I create a method called "preparedMediaSource" that return a MediaSource for ExoPlayer? Since if a user want to use the video in Resource/raw, he just need to override this method and using rawResourceDataSource instead (note: ExtractorMediaSource cannot get the video under res folder). Different MediaSource has different preparation.